src/sage/groups/abelian_gps/abelian_group.py

```diff
@@ -1301,10 +1301,27 @@ def __iter__(self):
             (1,)
             sage: list(G)
             [1]
+
+        We can also iterate over infinite groups::
+
+            sage: A = AbelianGroup([3,0,5,0])
+            sage: for a in A:
+            ....:   if a^2 == A([1, 2, 3, 4]):
+            ....:       print(a, a^2)
+            ....:       break
+            f0^2*f1*f2^4*f3^2 f0*f1^2*f2^3*f3^4
         """
         invs = self.gens_orders()
-        for t in mrange(invs):
-            yield self(t)
+        if 0 not in invs:
+            # The group is finite
+            for t in mrange(invs):
+                yield self(t)
+        else:
+            # A similar approach works for infinite groups.
+            # (This would also work for finite groups, but is more complicated.)
+            from sage.misc.mrange import cantor_product
+            for t in cantor_product(*[range(n) if n > 0 else ZZ for n in invs]):
+                yield self(t)
```

Comment:
I think that this can be simplified slightly:
```suggestion
        if 0 not in invs:
            # The group is finite
            yield from map(self, mrange(invs))
        else:
            # A similar approach works for infinite groups.
            # (This would also work for finite groups, but is more complicated.)
            from sage.misc.mrange import cantor_product
            yield from map(self, cantor_product(*[range(n) if n else ZZ for n in invs]))
```

I suppose that negative orders should not be allowed, although:
```
sage: G = AbelianGroup([-4])
sage: G.cardinality()
-4
```

Reply:
I agree that your code is an improvement, so I made the changes. (At least, I think I did, but I don't really understand github.)

Reply:
I opened issue #38967 to correct the problem with negative orders.